File: nothing/network.py
"""Network information and DNS management for Nothing phones."""

import re

from .device import run
from .exceptions import AdbError
from .models import DeviceInfo

# ---------------------------------------------------------------------------
# Known Private DNS aliases → real hostnames
# ---------------------------------------------------------------------------

_DNS_ALIASES: dict[str, str] = {
    "cloudflare": "one.one.one.one",
    "1.1.1.1":    "one.one.one.one",
    "adguard":    "dns.adguard.com",
    "google":     "dns.google",
    "quad9":      "dns.quad9.net",
}


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _shell(serial: str, cmd: str) -> str:
    """Run adb shell command; return stdout stripped, empty string on failure."""
    r = run(["adb", "-s", serial, "shell", cmd])
    return r.stdout.strip() if r.returncode == 0 else ""


def _setting(serial: str, namespace: str, key: str) -> str:
    """Read an Android setting via 'adb shell settings get'."""
    r = run(["adb", "-s", serial, "shell", "settings", "get", namespace, key])
    val = r.stdout.strip() if r.returncode == 0 else ""
    return "" if val in ("null", "null\n") else val


# ---------------------------------------------------------------------------
# 1. Network Info
# ---------------------------------------------------------------------------

def action_network_info(device: DeviceInfo) -> None:
    """Display network information for the connected Nothing phone."""
    s = device.serial

    # ── WiFi — use 'cmd wifi status' (stable across Android versions) ───────
    wifi_raw = _shell(s, "cmd wifi status")

    ssid       = ""
    bssid      = ""
    rssi       = ""
    link_speed = ""
    freq       = ""
    ip_wifi    = ""

    # Find the WifiInfo: line which contains all data in one line
    wifi_info_line = ""
    for line in wifi_raw.splitlines():
        if "WifiInfo:" in line:
            wifi_info_line = line
            break

    if wifi_info_line:
        m = re.search(r'SSID:\s*"([^"]*)"', wifi_info_line)
        if m:
            ssid = m.group(1)
        m = re.search(r'BSSID:\s*([0-9a-fA-F:]{17})', wifi_info_line)
        if m:
            bssid = m.group(1)
        m = re.search(r'RSSI:\s*(-?\d+)', wifi_info_line)
        if m:
            rssi = m.group(1)
        m = re.search(r'Link speed:\s*(\d+)', wifi_info_line)
        if m:
            link_speed = m.group(1)
        m = re.search(r'Frequency:\s*(\d+)', wifi_info_line)
        if m:
            freq = m.group(1)
        m = re.search(r'IP:\s*/([\d.]+)', wifi_info_line)
        if m:
            ip_wifi = m.group(1)

    # Fallback: ip addr show wlan0
    if not ip_wifi:
        wlan_raw = _shell(s, "ip addr show wlan0")
        m = re.search(r'inet\s+([\d.]+)/', wlan_raw)
        if m:
            ip_wifi = m.group(1)

    # Frequency band
    freq_band = ""
    if freq:
        try:
            f = int(freq)
            if f < 3000:
                freq_band = "2.4 GHz"
            elif f < 6000:
                freq_band = "5 GHz"
            else:
                freq_band = "6 GHz"
        except ValueError:
            pass

    # ── DNS ────────────────────────────────────────────────────────────────
    # Current DNS servers from /etc/resolv.conf or ndc resolver
    dns_raw = _shell(s, "getprop | grep -E 'net\\.dns[12]|dhcp.*dns'")
    dns_servers: list[str] = []
    for line in dns_raw.splitlines():
        m = re.search(r'net\.dns[12]\]:\s*\[([^\]]+)\]', line)
        if m and m.group(1).strip():
            dns_servers.append(m.group(1).strip())

    if not dns_servers:
        # Try ndc resolver dump
        ndc_raw = _shell(s, "ndc resolver getnetworkinfo 100")
        for line in ndc_raw.splitlines():
            m = re.search(r'DNS servers:\s*(.+)', line)
            if m:
                dns_servers = [d.strip() for d in m.group(1).split() if d.strip()]
                break

    # ── Private DNS ────────────────────────────────────────────────────────
    pdns_mode     = _setting(s, "global", "private_dns_mode")
    pdns_provider = _setting(s, "global", "private_dns_specifier")

    # ── Mobile network ─────────────────────────────────────────────────────
    operator    = _shell(s, "getprop gsm.operator.alpha")
    net_type    = _shell(s, "getprop gsm.network.type")

    # Multi-SIM: getprop may return comma-separated list — take the first non-empty
    if "," in operator:
        operator = next((o.strip() for o in operator.split(",") if o.strip()), "")
    if "," in net_type:
        net_type = next((n.strip() for n in net_type.split(",") if n.strip()), "")

    # ── Active connection type ─────────────────────────────────────────────
    conn_raw  = _shell(s, "dumpsys connectivity | grep -E 'NetworkAgentInfo.*CONNECTED|activeNetwork'")
    conn_type = "Unknown"
    if "WIFI" in conn_raw.upper():
        conn_type = "WiFi"
    elif "CELLULAR" in conn_raw.upper() or "MOBILE" in conn_raw.upper():
        conn_type = "Mobile"
    elif not conn_raw:
        conn_type = "None"

    # ── Output ────────────────────────────────────────────────────────────
    print(f"\n  Network Info — {device.model}\n")

    print(f"  {'Connection':<18}: {conn_type}")
    print()

    print(f"  WiFi")
    print(f"  {'  SSID':<18}: {ssid or '(not connected)'}")
    if ssid:
        print(f"  {'  BSSID':<18}: {bssid or 'n/a'}")
        print(f"  {'  Signal':<18}: {rssi + ' dBm' if rssi else 'n/a'}")
        print(f"  {'  Link Speed':<18}: {link_speed + ' Mbps' if link_speed else 'n/a'}")
        freq_display = f"{freq} MHz ({freq_band})" if freq and freq_band else (freq or "n/a")
        print(f"  {'  Frequency':<18}: {freq_display}")
        print(f"  {'  IP Address':<18}: {ip_wifi or 'n/a'}")
    print()

    print(f"  DNS")
    if dns_servers:
        for i, srv in enumerate(dns_servers, 1):
            print(f"  {'  Server ' + str(i):<18}: {srv}")
    else:
        print(f"  {'  Servers':<18}: (not available)")

    pdns_display = pdns_mode or "off"
    if pdns_mode == "hostname" and pdns_provider:
        pdns_display = f"hostname ({pdns_provider})"
    elif pdns_mode == "opportunistic":
        pdns_display = "automatic"
    print(f"  {'  Private DNS':<18}: {pdns_display}")
    print()

    print(f"  Mobile")
    print(f"  {'  Operator':<18}: {operator or 'n/a'}")
    print(f"  {'  Network Type':<18}: {net_type or 'n/a'}")
    print()


# ---------------------------------------------------------------------------
# 2. DNS Set
# ---------------------------------------------------------------------------

def action_dns_set(device: DeviceInfo, provider: str | None) -> None:
    """Set or display the Private DNS configuration."""
    s = device.serial

    # ── Read-only mode ─────────────────────────────────────────────────────
    if provider is None:
        mode     = _setting(s, "global", "private_dns_mode")
        specifier = _setting(s, "global", "private_dns_specifier")

        print(f"\n  Private DNS — {device.model}\n")
        mode_display = mode or "off"
        if mode == "hostname":
            mode_display = f"hostname"
        elif mode == "opportunistic":
            mode_display = "automatic (opportunistic)"
        print(f"  {'Mode':<12}: {mode_display}")
        if mode == "hostname":
            print(f"  {'Provider':<12}: {specifier or '(not set)'}")
        print()
        return

    # ── Disable ────────────────────────────────────────────────────────────
    if provider.lower() == "off":
        r = run(["adb", "-s", s, "shell", "settings", "put", "global",
                 "private_dns_mode", "off"])
        if r.returncode != 0:
            raise AdbError(f"Failed to disable Private DNS: {r.stderr.strip()}")
        print(f"  Private DNS disabled on {device.model}.")
        return

    # ── Resolve alias ──────────────────────────────────────────────────────
    hostname = _DNS_ALIASES.get(provider.lower(), provider)

    # ── Set hostname mode ──────────────────────────────────────────────────
    r1 = run(["adb", "-s", s, "shell", "settings", "put", "global",
              "private_dns_specifier", hostname])
    r2 = run(["adb", "-s", s, "shell", "settings", "put", "global",
              "private_dns_mode", "hostname"])

    if r1.returncode != 0 or r2.returncode != 0:
        raise AdbError(
            f"Failed to set Private DNS: "
            f"{(r1.stderr or r2.stderr).strip()}"
        )

    alias_note = f" (alias for {hostname})" if hostname != provider.lower() else ""
    print(f"  Private DNS set to '{hostname}'{alias_note} on {device.model}.")


# ---------------------------------------------------------------------------
# 3. Port Forwarding
# ---------------------------------------------------------------------------

def action_port_forward(
    device: DeviceInfo,
    local: str | None,
    remote: str | None,
    clear: bool,
) -> None:
    """Manage ADB port forwards for the connected Nothing phone."""
    s = device.serial

    # ── Remove all ────────────────────────────────────────────────────────
    if clear:
        r = run(["adb", "-s", s, "forward", "--remove-all"])
        if r.returncode != 0:
            raise AdbError(f"Failed to remove forwards: {r.stderr.strip()}")
        print(f"  All port forwards removed on {device.model}.")
        return

    # ── Add new forward ───────────────────────────────────────────────────
    if local is not None and remote is not None:
        local_spec  = f"tcp:{local}"
        remote_spec = f"tcp:{remote}"
        r = run(["adb", "-s", s, "forward", local_spec, remote_spec])
        if r.returncode != 0:
            raise AdbError(
                f"Failed to create forward {local_spec} -> {remote_spec}: "
                f"{r.stderr.strip()}"
            )
        print(f"  Forward added: {local_spec} -> {remote_spec} on {device.model}.")
        return

    # ── List active forwards ──────────────────────────────────────────────
    r_fwd = run(["adb", "-s", s, "forward", "--list"])
    r_rev = run(["adb", "-s", s, "reverse", "--list"])

    fwd_lines = [l.strip() for l in r_fwd.stdout.splitlines() if l.strip()] \
        if r_fwd.returncode == 0 else []
    rev_lines = [l.strip() for l in r_rev.stdout.splitlines() if l.strip()] \
        if r_rev.returncode == 0 else []

    print(f"\n  Port Forwards — {device.model}\n")

    print(f"  Forwards (host -> device):")
    if fwd_lines:
        for line in fwd_lines:
            # Format: <serial> <local> <remote>
            parts = line.split()
            if len(parts) >= 3:
                print(f"    {parts[1]}  ->  {parts[2]}")
            else:
                print(f"    {line}")
    else:
        print("    (none)")

    print()
    print(f"  Reverse forwards (device -> host):")
    if rev_lines:
        for line in rev_lines:
            parts = line.split()
            if len(parts) >= 3:
                print(f"    {parts[1]}  ->  {parts[2]}")
            else:
                print(f"    {line}")
    else:
        print("    (none)")

    print()

File: go/internal/network/network.go
// Package network provides network info, DNS, and port-forwarding management.
package network

import (
	"fmt"
	"regexp"
	"strings"
	"sync"

	"github.com/Limplom/nothingctl/internal/adb"
	nterrors "github.com/Limplom/nothingctl/internal/errors"
)

// DNS aliases → real hostnames
var dnsAliases = map[string]string{
	"cloudflare": "one.one.one.one",
	"1.1.1.1":    "one.one.one.one",
	"adguard":    "dns.adguard.com",
	"google":     "dns.google",
	"quad9":      "dns.quad9.net",
}

var ssidRe      = regexp.MustCompile(`SSID:\s*"([^"]*)"`)
var bssidRe     = regexp.MustCompile(`BSSID:\s*([0-9a-fA-F:]{17})`)
var rssiRe      = regexp.MustCompile(`RSSI:\s*(-?\d+)`)
var linkSpeedRe = regexp.MustCompile(`Link speed:\s*(\d+)`)
var freqRe      = regexp.MustCompile(`Frequency:\s*(\d+)`)
var ipWifiRe    = regexp.MustCompile(`IP:\s*/([\d.]+)`)
var inetRe      = regexp.MustCompile(`inet\s+([\d.]+)/`)
var dns12Re     = regexp.MustCompile(`net\.dns[12]\]:\s*\[([^\]]+)\]`)
var ndcDNSRe    = regexp.MustCompile(`DNS servers:\s*(.+)`)

// ActionNetworkInfo displays network information.
func ActionNetworkInfo(serial, model string) error {
	var wifiRaw, dnsRaw, operator, netType, connRaw string
	var pdnsMode, pdnsProvider string

	var wg sync.WaitGroup
	wg.Add(6)
	go func() { defer wg.Done(); wifiRaw = adb.ShellStr(serial, "cmd wifi status") }()
	go func() {
		defer wg.Done()
		dnsRaw = adb.ShellStr(serial, "getprop | grep -E 'net\\.dns[12]|dhcp.*dns'")
	}()
	go func() { defer wg.Done(); operator = adb.ShellStr(serial, "getprop gsm.operator.alpha") }()
	go func() { defer wg.Done(); netType = adb.ShellStr(serial, "getprop gsm.network.type") }()
	go func() {
		defer wg.Done()
		connRaw = adb.ShellStr(serial, "dumpsys connectivity | grep -E 'NetworkAgentInfo.*CONNECTED|activeNetwork'")
	}()
	go func() {
		defer wg.Done()
		pdnsMode = adb.Setting(serial, "global", "private_dns_mode")
		pdnsProvider = adb.Setting(serial, "global", "private_dns_specifier")
	}()
	wg.Wait()

	var ssid, bssid, rssi, linkSpeed, freq, ipWifi string

	// Find WifiInfo: line
	for _, line := range strings.Split(wifiRaw, "\n") {
		line = strings.TrimRight(line, "\r")
		if !strings.Contains(line, "WifiInfo:") {
			continue
		}
		if m := ssidRe.FindStringSubmatch(line); m != nil {
			ssid = m[1]
		}
		if m := bssidRe.FindStringSubmatch(line); m != nil {
			bssid = m[1]
		}
		if m := rssiRe.FindStringSubmatch(line); m != nil {
			rssi = m[1]
		}
		if m := linkSpeedRe.FindStringSubmatch(line); m != nil {
			linkSpeed = m[1]
		}
		if m := freqRe.FindStringSubmatch(line); m != nil {
			freq = m[1]
		}
		if m := ipWifiRe.FindStringSubmatch(line); m != nil {
			ipWifi = m[1]
		}
		break
	}

	// Fallback IP from ip addr show wlan0
	if ipWifi == "" {
		wlanRaw := adb.ShellStr(serial, "ip addr show wlan0")
		if m := inetRe.FindStringSubmatch(wlanRaw); m != nil {
			ipWifi = m[1]
		}
	}

	// Frequency band
	freqBand := ""
	if freq != "" {
		var f int
		fmt.Sscanf(freq, "%d", &f)
		switch {
		case f < 3000:
			freqBand = "2.4 GHz"
		case f < 6000:
			freqBand = "5 GHz"
		default:
			freqBand = "6 GHz"
		}
	}

	// DNS servers
	var dnsServers []string
	for _, line := range strings.Split(dnsRaw, "\n") {
		line = strings.TrimRight(line, "\r")
		if m := dns12Re.FindStringSubmatch(line); m != nil {
			if v := strings.TrimSpace(m[1]); v != "" {
				dnsServers = append(dnsServers, v)
			}
		}
	}
	if len(dnsServers) == 0 {
		ndcRaw := adb.ShellStr(serial, "ndc resolver getnetworkinfo 100")
		for _, line := range strings.Split(ndcRaw, "\n") {
			line = strings.TrimRight(line, "\r")
			if m := ndcDNSRe.FindStringSubmatch(line); m != nil {
				for _, d := range strings.Fields(m[1]) {
					if d != "" {
						dnsServers = append(dnsServers, d)
					}
				}
				break
			}
		}
	}

	// Handle comma-separated multi-SIM values
	if strings.Contains(operator, ",") {
		for _, o := range strings.Split(operator, ",") {
			if strings.TrimSpace(o) != "" {
				operator = strings.TrimSpace(o)
				break
			}
		}
	}
	if strings.Contains(netType, ",") {
		for _, n := range strings.Split(netType, ",") {
			if strings.TrimSpace(n) != "" {
				netType = strings.TrimSpace(n)
				break
			}
		}
	}

	// Active connection type
	connType := "Unknown"
	upper := strings.ToUpper(connRaw)
	switch {
	case strings.Contains(upper, "WIFI"):
		connType = "WiFi"
	case strings.Contains(upper, "CELLULAR") || strings.Contains(upper, "MOBILE"):
		connType = "Mobile"
	case connRaw == "":
		connType = "None"
	}

	// Output
	fmt.Printf("\n  Network Info \u2014 %s\n\n", model)
	fmt.Printf("  %-18s: %s\n", "Connection", connType)
	fmt.Println()

	fmt.Println("  WiFi")
	if ssid == "" {
		fmt.Printf("  %-18s: %s\n", "  SSID", "(not connected)")
	} else {
		fmt.Printf("  %-18s: %s\n", "  SSID", ssid)
		bssidDisplay := bssid
		if bssidDisplay == "" {
			bssidDisplay = "n/a"
		}
		fmt.Printf("  %-18s: %s\n", "  BSSID", bssidDisplay)
		if rssi != "" {
			fmt.Printf("  %-18s: %s dBm\n", "  Signal", rssi)
		} else {
			fmt.Printf("  %-18s: n/a\n", "  Signal")
		}
		if linkSpeed != "" {
			fmt.Printf("  %-18s: %s Mbps\n", "  Link Speed", linkSpeed)
		} else {
			fmt.Printf("  %-18s: n/a\n", "  Link Speed")
		}
		freqDisplay := "n/a"
		if freq != "" && freqBand != "" {
			freqDisplay = fmt.Sprintf("%s MHz (%s)", freq, freqBand)
		} else if freq != "" {
			freqDisplay = freq
		}
		fmt.Printf("  %-18s: %s\n", "  Frequency", freqDisplay)
		ipDisplay := ipWifi
		if ipDisplay == "" {
			ipDisplay = "n/a"
		}
		fmt.Printf("  %-18s: %s\n", "  IP Address", ipDisplay)
	}
	fmt.Println()

	fmt.Println("  DNS")
	if len(dnsServers) > 0 {
		for i, srv := range dnsServers {
			fmt.Printf("  %-18s: %s\n", fmt.Sprintf("  Server %d", i+1), srv)
		}
	} else {
		fmt.Printf("  %-18s: (not available)\n", "  Servers")
	}

	pdnsDisplay := "off"
	if pdnsMode != "" {
		pdnsDisplay = pdnsMode
	}
	if pdnsMode == "hostname" && pdnsProvider != "" {
		pdnsDisplay = fmt.Sprintf("hostname (%s)", pdnsProvider)
	} else if pdnsMode == "opportunistic" {
		pdnsDisplay = "automatic"
	}
	fmt.Printf("  %-18s: %s\n", "  Private DNS", pdnsDisplay)
	fmt.Println()

	fmt.Println("  Mobile")
	opDisplay := operator
	if opDisplay == "" {
		opDisplay = "n/a"
	}
	ntDisplay := netType
	if ntDisplay == "" {
		ntDisplay = "n/a"
	}
	fmt.Printf("  %-18s: %s\n", "  Operator", opDisplay)
	fmt.Printf("  %-18s: %s\n", "  Network Type", ntDisplay)
	fmt.Println()
	return nil
}

// ActionDNSSet sets or displays Private DNS configuration.
func ActionDNSSet(serial, model, provider string) error {
	if provider == "" {
		// Read-only mode
		mode := adb.Setting(serial, "global", "private_dns_mode")
		specifier := adb.Setting(serial, "global", "private_dns_specifier")

		fmt.Printf("\n  Private DNS \u2014 %s\n\n", model)
		modeDisplay := "off"
		if mode != "" {
			modeDisplay = mode
		}
		if mode == "opportunistic" {
			modeDisplay = "automatic (opportunistic)"
		}
		fmt.Printf("  %-12s: %s\n", "Mode", modeDisplay)
		if mode == "hostname" {
			sp := specifier
			if sp == "" {
				sp = "(not set)"
			}
			fmt.Printf("  %-12s: %s\n", "Provider", sp)
		}
		fmt.Println()
		return nil
	}

	if strings.ToLower(provider) == "off" {
		_, stderr, code := adb.Run([]string{"adb", "-s", serial, "shell",
			"settings", "put", "global", "private_dns_mode", "off"})
		if code != 0 {
			return nterrors.AdbError(fmt.Sprintf("Failed to disable Private DNS: %s", strings.TrimSpace(stderr)))
		}
		fmt.Printf("  Private DNS disabled on %s.\n", model)
		return nil
	}

	hostname := provider
	if alias, ok := dnsAliases[strings.ToLower(provider)]; ok {
		hostname = alias
	}

	_, stderr1, code1 := adb.Run([]string{"adb", "-s", serial, "shell",
		"settings", "put", "global", "private_dns_specifier", hostname})
	_, stderr2, code2 := adb.Run([]string{"adb", "-s", serial, "shell",
		"settings", "put", "global", "private_dns_mode", "hostname"})

	if code1 != 0 || code2 != 0 {
		errMsg := strings.TrimSpace(stderr1 + stderr2)
		return nterrors.AdbError(fmt.Sprintf("Failed to set Private DNS: %s", errMsg))
	}

	aliasNote := ""
	if hostname != strings.ToLower(provider) {
		aliasNote = fmt.Sprintf(" (alias for %s)", hostname)
	}
	fmt.Printf("  Private DNS set to '%s'%s on %s.\n", hostname, aliasNote, model)
	return nil
}

// ActionPortForward manages ADB port forwards.
// local and remote are port numbers as strings (e.g. "8080"). Pass empty strings for list mode.
func ActionPortForward(serial, model, local, remote string, clear bool) error {
	if clear {
		_, stderr, code := adb.Run([]string{"adb", "-s", serial, "forward", "--remove-all"})
		if code != 0 {
			return nterrors.AdbError(fmt.Sprintf("Failed to remove forwards: %s", strings.TrimSpace(stderr)))
		}
		fmt.Printf("  All port forwards removed on %s.\n", model)
		return nil
	}

	if local != "" && remote != "" {
		localSpec := "tcp:" + local
		remoteSpec := "tcp:" + remote
		_, stderr, code := adb.Run([]string{"adb", "-s", serial, "forward", localSpec, remoteSpec})
		if code != 0 {
			return nterrors.AdbError(fmt.Sprintf("Failed to create forward %s -> %s: %s",
				localSpec, remoteSpec, strings.TrimSpace(stderr)))
		}
		fmt.Printf("  Forward added: %s -> %s on %s.\n", localSpec, remoteSpec, model)
		return nil
	}

	fwdOut, _, fwdCode := adb.Run([]string{"adb", "-s", serial, "forward", "--list"})
	revOut, _, revCode := adb.Run([]string{"adb", "-s", serial, "reverse", "--list"})

	fmt.Printf("\n  Port Forwards \u2014 %s\n\n", model)

	fmt.Println("  Forwards (host -> device):")
	if fwdCode == 0 {
		lines := filterNonEmpty(strings.Split(fwdOut, "\n"))
		if len(lines) > 0 {
			for _, line := range lines {
				parts := strings.Fields(line)
				if len(parts) >= 3 {
					fmt.Printf("    %s  ->  %s\n", parts[1], parts[2])
				} else {
					fmt.Printf("    %s\n", line)
				}
			}
		} else {
			fmt.Println("    (none)")
		}
	} else {
		fmt.Println("    (none)")
	}

	fmt.Println()
	fmt.Println("  Reverse forwards (device -> host):")
	if revCode == 0 {
		lines := filterNonEmpty(strings.Split(revOut, "\n"))
		if len(lines) > 0 {
			for _, line := range lines {
				parts := strings.Fields(line)
				if len(parts) >= 3 {
					fmt.Printf("    %s  ->  %s\n", parts[1], parts[2])
				} else {
					fmt.Printf("    %s\n", line)
				}
			}
		} else {
			fmt.Println("    (none)")
		}
	} else {
		fmt.Println("    (none)")
	}

	fmt.Println()
	return nil
}

func filterNonEmpty(lines []string) []string {
	var out []string
	for _, l := range lines {
		l = strings.TrimRight(l, "\r")
		if strings.TrimSpace(l) != "" {
			out = append(out, l)
		}
	}
	return out
}
